pkg/base-dev/lib/pill.hoon

```diff
@@ -128,4 +142,225 @@
+  ::
+  |=  [sys=path dez=(list [desk path]) prime=?]
+  ^-  pill
+  =/  bas=path  (scag 3 sys)
+  ::  compiler-source: hoon source file producing compiler, `sys/hoon`
+  ::
+  =+  compiler-source=.^(@t %cx (welp sys /hoon/hoon))
+  ::
+  ::  compiler-twig: compiler as hoon expression
+  ::
+  ~&  %brass-parsing
+  =+  compiler-twig=(rain /sys/hoon/hoon compiler-source)
+  ~&  %brass-parsed
+  ::
+  ::  compiler-formula: compiler as nock formula
+  ::
+  ~&  %brass-compiling
+  =+  compiler-formula=q:(~(mint ut %noun) %noun compiler-twig)
+  ~&  %brass-compiled
+  ::
+  ::  arvo-source: hoon source file producing arvo kernel, `sys/arvo`
+  ::
+  =+  arvo-source=.^(@t %cx (welp sys /arvo/hoon))
+  ::
+  ::  boot-ova: startup events
+  ::
+  =/  boot-ova=(list)
+    :~  aeon:eden:part
+        boot:eden:part
+        compiler-formula
+        compiler-source
+        arvo-source
+    ==
+  ::  a pill is a 3-tuple of event-lists: [boot kernel userspace]
+  ::
+  :+  %pill  %brass
+  :+  boot-ova
+    :~  (boot-ovum:pill compiler-source arvo-source)
+        (file-ovum2:pill bas)
+    ==
+  =.  dez  (snoc dez [%base bas])
+  %+  weld
+    (turn dez file-ovum)
+  ?.  prime  ~
+  [(prep-ovum (turn dez tail))]~
+::
+++  ivory
+  |=  sys=path
+  ^-  pill
+  =/  lib  (snoc (scag 3 sys) %lib)
+  |^  =/  ver
+        =/  sub  *(trap vase)
+        =.  sub  (build-sys sub %hoon)
+        =.  sub  (build-sys sub %arvo)
+        =.  sub  (build-sys sub %lull)
+        =.  sub  (build-sys sub %zuse)
+        =.  sub  (build-lib sub & %ethereum)
+        =.  sub  (build-lib sub & %azimuth)
+        (build-lib sub | %vere)
+      =/  nok  !.
+        =>  *[ver=(trap vase) ~]
+        !=  q:$:ver
+      ivory/[nok ver ~]
+  ::
+  ++  build-sys
+    |=  [sub=(trap vase) nam=term]  ^-  (trap vase)
+    ~>  %slog.[0 leaf+"ivory: building /sys/{(trip nam)}"]
+    (swat sub (rain /sys/[nam]/hoon .^(@t cx+(welp sys /[nam]/hoon))))
+  ::
+  ++  build-lib
+    |=  [sub=(trap vase) imp=? nam=term]  ^-  (trap vase)
+    ~>  %slog.[0 leaf+"ivory: building /lib/{(trip nam)}"]
+    =/  hun=hoon
+      %+  mist  /lib/[nam]/hoon
+      .^(@t cx+(welp lib /[nam]/hoon))
+    ?.  imp  (swat sub hun)
+    (swel sub [%ktts nam hun])
+  ::  +mist: +rain but skipping past ford runes
+  ::
+  ++  mist
+    |=  [bon=path txt=@]
+    ^-  hoon
+    =+  vas=vast
+    ~|  bon
+    %+  scan  (trip txt)
+    %-  full
+    =;  fud
+      (ifix [;~(plug gay fud) gay] tall:vas(wer bon))
+    %-  star
+    ;~  pose  vul
+      %+  ifix  [fas (just `@`10)]
+      (star ;~(less (just `@`10) next))
+    ==
+  ::  +swel: +swat but with +slop
+  ::
+  ++  swel
+    |=  [tap=(trap vase) gen=hoon]
+    ^-  (trap vase)
+    =/  gun  (~(mint ut p:$:tap) %noun gen)
+    =>  [tap=tap gun=gun]
+    |.  ~+
+    =/  pro  q:$:tap
+    [[%cell p.gun p:$:tap] [.*(pro q.gun) pro]]
+  --
+::
+++  events
+  |%
+  +$  prop  [%prop meta tier (list ovum)]
+  +$  meta  term
+  +$  tier  ?(%pre-userspace %post-userspace)
```

Comment:
I'm tempted to say that these types should be in the kernel somewhere. They are being referenced in the ipc boot message specification. But I'm not sure where.

I don't like the aesthetic of `$tier`, I'd rather it be something like `?(%fore %hind)`. And I'm having trouble coming up with a use-case for other tiers ("pre-kernelspace"?). Thoughts?